### PR TITLE
Support Externally Created virtual functions

### PR DESCRIPTION
**Pull request by SchSeba**
this PR adds the new API to allow the operator to use virtual functions
that the user configures manually on the system

this will allow new use-cases like using a VF as the SDN primary nic.
the nic will be created before the operator starts by the user, for
example using nmstate or NetworkManager and the operator will allocate
all the remaining vfs to pods.

### CODE REVIEW

File: pkg/utils/utils.go
Comment (zeeke) [inline pkg/utils/utils.go:220]: nit:

`SyncNodeState(): PF name %s with pci address %s has VFs configured but they weren't created by the sriov operator. Skipping the device reset`

**Review by zeeke — COMMENTED** [review]
Overall, PR looks good to me. I added a few comments error messages

**Comment by zeeke** [comment]
LGTM

File: api/v1/sriovnetworknodepolicy_types.go
Comment (adrianchiris) [inline api/v1/sriovnetworknodepolicy_types.go:63]: how does ExternallyCreated relate to other parameters here ?

we need to have this documented so users will know what to expect when using it.

File: api/v1/sriovnetworknodestate_types.go
Comment (SchSeba) [reply api/v1/sriovnetworknodestate_types.go:36]: I update the cases in the validation webhook for everyone of them.
right now and until there is a use-case we will not allow ExternallyCreated + switchdev mode

Comment (adrianchiris) [reply api/v1/sriovnetworknodestate_types.go:36]: isnt the use-case you are aiming for is for ovn-kubernetes + HW offload ?
is that only for dpu or also regular nics ?

File: api/v1/sriovnetworknodepolicy_types.go
Comment (SchSeba) [reply api/v1/sriovnetworknodepolicy_types.go:63]: I add documentation for it let me know if you need anything else :)

Comment (moshe010) [reply api/v1/sriovnetworknodepolicy_types.go:63]: is there a reason to do it per interface and not global? 
Do you have use-case when some SR-IOV will be create by SR-IOV opeartor and some not?

Comment (moshe010) [reply api/v1/sriovnetworknodepolicy_types.go:63]: I read it now in the description. You want to use it for  CNI primary network and secondary will VF still be managed by sriov operator

File: pkg/webhook/validate.go
Comment (adrianchiris) [inline pkg/webhook/validate.go:171]: why do you want to block it ?

if its externally created and device is already in switchdev mode then its OK no ?

File: pkg/utils/host.go
Comment (adrianchiris) [inline pkg/utils/host.go:46]: you can call MkdirAll[1] on PfAppliedConfg

[1[ https://pkg.go.dev/os#MkdirAll

File: pkg/utils/utils.go
Comment (adrianchiris) [inline pkg/utils/utils.go:323]: why can the admin mac be "externally configured" ? that should be part of sriov configuration performed by the external service.

File: pkg/utils/host.go
Comment (adrianchiris) [reply pkg/utils/host.go:112]: A: Because SyncNodeState is called with chroot

File: pkg/utils/utils.go
Comment (adrianchiris) [inline pkg/utils/utils.go:220]: if entry was not found, does this nessecarily mean that VFs were not created by sriov-operator ?

**Review by adrianchiris — COMMENTED** [review]
starting to wonder if we need a design doc for this one.

it feels that we need to get an agreement on what exactly ExternallyManaged means
then outline the changes in the different flows of sriov-network-operator.

that way:
1. we agree on what we want to achieve 
2. make sure this PR is doing what we want

**Comment by adrianchiris** [comment]
also UT need to be added where possible.

File: pkg/webhook/validate.go
Comment (SchSeba) [reply pkg/webhook/validate.go:171]: we should block right now because all the systemd and bash script we do for this will be hard to manage.
and right know I am not aware of any request for this.

File: pkg/daemon/daemon.go
Comment (SchSeba) [reply pkg/daemon/daemon.go:471]: because we may get a new spec that we don't expect.

File: pkg/daemon/daemon.go
Comment (SchSeba) [reply pkg/daemon/daemon.go:477]: I agree but I will like to do it in a followup PR if possible

File: pkg/daemon/daemon.go
Comment (SchSeba) [reply pkg/daemon/daemon.go:476]: sure we can go with that

**Comment by SchSeba** [comment]
I already add UT to the webhook folder.
for the other stuff it's a bit harder with how the config-daemon is configured right now I think

File: README.md
Comment (SchSeba) [reply README.md:217]: That was my initial idea but the problem is that there is no way using nmstate or network manager to tell what is the driver you will like for every PF that is the reason I decide to let the operator does it

File: README.md
Comment (e0ne) [reply README.md:216]: +1 for `ExternallyManaged` name

File: pkg/utils/host.go
Comment (e0ne) [inline pkg/utils/host.go:42]: nit: IMO`EnsureOperatorConfigFolder` is more clear name

File: pkg/utils/host.go
Comment (e0ne) [inline pkg/utils/host.go:15]: It's duplicated with https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/37ddcd4fd3c011241348bd6a3efce9d426b07f6a/pkg/utils/sriov.go#L31

File: pkg/utils/host.go
Comment (e0ne) [inline pkg/utils/host.go:30]: Why don't we store SriovNetworkNodeState object here like we did for systemd service?

**Review by e0ne — CHANGES_REQUESTED** [review]

File: pkg/utils/host.go
Comment (SchSeba) [reply pkg/utils/host.go]: the systemd file is different.
here we have a per PF info of the latest applied policy so we know if we need to reset the device or not

**Review by zeeke — COMMENTED** [review]
Left some nit. It would be great if we could make some unit test around daemon/generic-plugin logic

File: pkg/utils/store.go
Comment (SchSeba) [reply pkg/utils/store.go]: sure done

File: pkg/utils/store.go
Comment (SchSeba) [reply pkg/utils/store.go:133]: you are right I miss the returns! thanks!

**Comment by SchSeba** [comment]
> Left some nit. It would be great if we could make some unit test around daemon/generic-plugin logic

I am working on a separate PR after we move the drain to a controller a refactor in the daemon that will use more interfaces so we can mock easily

**Review by zeeke — COMMENTED** [review]
Left a few minor comments. Overall shape looks good to me.

File: pkg/webhook/validate.go
Comment (SchSeba) [reply pkg/webhook/validate.go:307]: the copy past issue :P

File: pkg/utils/store.go
Comment (SchSeba) [reply pkg/utils/store.go:30]: good point but i wanted to be the same as the host package.

maybe I can change both of them in a separate PR

File: pkg/utils/store.go
Comment (SchSeba) [reply pkg/utils/store.go:85]: done

File: pkg/utils/store.go
Comment (SchSeba) [reply pkg/utils/store.go]: right!

**Comment by SchSeba** [comment]
Hi @zeeke let me know if the changes are good and I will rebase the commit messages

**Review by zeeke — APPROVED** [review]
LGTM
Thanks for addressing my comments

File: pkg/utils/utils.go
Comment (e0ne) [inline pkg/utils/utils.go:933]: Do we want to log stderr here?

File: pkg/utils/utils.go
Comment (e0ne) [inline pkg/utils/utils.go:945]: Please, move this template to constant to make it more readable

File: pkg/utils/utils.go
Comment (SchSeba) [reply pkg/utils/utils.go:935]: done

**Review by e0ne — APPROVED** [review]
LGTM. Thanks for addressing my comments

**Comment by SchSeba** [comment]
Thanks folks!

Merging after 2 maintainers approval

**Pull request merged**
